3p/ampcontext-integration.js

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {AbstractAmpContext} from './ampcontext';
+import {computeInMasterFrame} from './3p';
+import {dev, user} from '../src/log';
+
+
+/**
+ * Returns the "master frame" for all widgets of a given type.
+ * This frame should be used to e.g. fetch scripts that can
+ * be reused across frames.
+ * once experiment is removed.
+ * @param {!Window} win
+ * @param {string} type
+ * @return {!Window}
+ */
+// DUPLICATE IN 3p/integration
+function masterSelection(win, type) {
+  // The master has a special name.
+  const masterName = 'frame_' + type + '_master';
+  let master;
+  try {
+    // Try to get the master from the parent. If it does not
+    // exist yet we get a security exception that we catch
+    // and ignore.
+    master = win.parent.frames[masterName];
+  } catch (expected) {
+    /* ignore */
+  }
+  if (!master) {
+    // No master yet, rename ourselves to be master. Yaihh.
+    win.name = masterName;
+    master = win;
+  }
+  return master;
+}
+
+
+export class IntegrationAmpContext extends AbstractAmpContext {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    super(win);
+
+    /** @private {?string} */
+    this.embedType_ = null;
+
+    /** @private {boolean} */
+    this.updateDimensionsEnabled_ = false;
+  }
+
+  /** @override */
+  isAbstractImplementation_() {
+    return false;
+  }
+
+  /** @return {!Window} */
+  get master() {
+    return this.master_();
+  }
+
+  /** @return {!Window} */
+  master_() {
+    return masterSelection(this.win_, dev().assertString(this.embedType_));
+  }
+
+  /** @return {boolean} */
+  get isMaster() {
+    return this.isMaster_();
+  }
+
+  /** @return {boolean} */
+  isMaster_() {
+    return this.master == this.win_;
+  }
+
+  /** @param {string} type */
+  setEmbedType(type) {
+    this.embedType_ = type;
+
+    if (type === 'facebook' || type === 'twitter') {
+      // Only make this available to selected embeds until the
+      // generic solution is available.
+      this.updateDimensionsEnabled_ = true;
+    }
+  }
+
+  /**
+   * @param {number} width
+   * @param {number} height
+   */
+  updateDimensions(width, height) {
+    dev().assert(this.updateDimensionsEnabled_, 'Not available.');
+    this.requestResize(width, height);
+  }
+
+  notifyBootstrapLoaded() {
```

Comment:
to be consistent, you can call this method `bootsrapLoaded`, or you change the following method `renderStart` to `notifyRenderStart`